### DISCUSSION TITLE
"audio_info: slow stream, dropouts are possible", would PSRAM help?

### DISCUSSION THREAD

**Discussion by probonopd** [schreibfaul1/ESP32-audioI2S#486] [Q&A]
I am trying to play https://audio.hbr.org/ideacast/20230206141810-896_WhyManyCompaniesGetLayoffsWrong.mp3, but I get stuttering playback. Actually the URL is just an example, I have observed this with other MP3 URLs too.

My computer can download this at 3.57MB/s, so I think the

```
audio_info: slow stream, dropouts are possible
audio_info: slow stream, dropouts are possible
audio_info: slow stream, dropouts are possible
```

is not due to a slow server.

Would PSRAM help for this URL?

Or is the synchronous HTTP stack the actual bottleneck here?

`-D FREERTOS_HZ=240` did not make a difference here.

Unfortunately I currently don't have any ESP32 with PSRAM. So I'd appreciate if someone could test whether they can play the mp3 above without stuttering.

Thanks!

**Comment by pdurys** [comment]
I have ESP32-S3 inside LilyGO T-Display-S3
It has PSRAM and plays your mp3 file just fine.

But could you check something else, please?
In my research over last week, I am still trying to understand what is going on.

In my project I have also OneButton library to simplify buttons handling. I am also using lvgl library.
Buttons in my board are connected to GPIO 0 and 14.
Most of the time when I compile code in PIO and upload I have constant GPIO 0 LOW, like it would be pressed forever.
This causes also ```slow stream, dropouts are possible``` but that might be just reaction of audioI2S library to overload.
I have no issues with slow stream, dropouts are possible while compiling under Arduino IDE.

**Reply by probonopd** [reply]
Thank you very much @pdurys. Could it be that you have slightly different versions of the ESP32 core/compiler/libraries in the Arduino IDE vs. PlatformIO?

**Reply by pdurys** [reply]
It might be, but just cannot find out what is affecting that GPIO 0 always LOW signal.
If I find something I will share it here.

**Comment by schreibfaul1** [comment]
Interesting, then uses the Arduino IDE choose a different Arduino or IDF version.
Here's how you can find out:
````c++
    uint8_t avMajor  = ESP_ARDUINO_VERSION_MAJOR;
    uint8_t avMinor  = ESP_ARDUINO_VERSION_MINOR;
    uint8_t avPatch  = ESP_ARDUINO_VERSION_PATCH;
    Serial.printf("ESP32 Chip: %s\n", chipModel);
    Serial.printf("Arduino Version: %d.%d.%d\n", avMajor, avMinor, avPatch);
    uint8_t idfMajor = ESP_IDF_VERSION_MAJOR;
    uint8_t idfMinor = ESP_IDF_VERSION_MINOR;
    uint8_t idfPatch = ESP_IDF_VERSION_PATCH;
    Serial.printf("ESP-IDF Version: %d.%d.%d\n", idfMajor, idfMinor, idfPatch);
````
The concept of the input buffers resembles that of a leaky bucket, the amount that drains is predetermined (nr. of channels, samplerate...). At the beginning the bucket is filled. If more is flowing out than is flowing in, the fill level decreases. If the bucket is almost empty, the message appears: slow stream, dropouts are possible. Increasing the buffer does not make sense. The bottleneck is the WiFiClient and WiFiClientSecure.

**Reply by probonopd** [reply]
Well, if the water is not flowing in at a constant rate but in chunks, then a larger bucket can make sense. And isn't having PSRAM "a larger bucket"?

**Comment by pdurys** [comment]
Arduino Version 2.0.5 - going to see if there will be any change with latest 2.0.6
ESP-IDF Version 4.4.2

My WiFi, WiFiClientSecure and HTTPClient libs are exactly the same in both Projects.
There are minor differences in lvgl library.

**Reply by pdurys** [reply]
Just have checked both OK and NOK projects with latest ESP Arduino 2.0.6 and ESP-IDF 4.4.3 The same observations.

More details
OK lvgl @ 8.3.0-dev and ESP32-audioI2S-master @ 2.0.0
NOK lvgl @ 8.3.4 and ESP32-audioI2S @ 2.1.0

If that helps here you are log from Terminal

NOK. Please note this error audio_info:            MP3 decode error -2 : MAINDATA_UNDERFLOW
Maybe this will tell you something.

```
audio_info:            PSRAM found, inputBufferSize: 283615 bytes
audio_info:            buffers freed, free Heap: 80872 bytes
audio_info:            Connect to new host: "https://stream.rcs.revma.com/an1ugyygzk8uv"
audio_info:            Connect to "stream.rcs.revma.com" on port 443, extension "/an1ugyygzk8uv"
audio_info:            SSL has been established in 645 ms, free Heap: 34556 bytes
Setup completed.
audio_info:            HTTP/1.0 302 Found
audio_info:            redirect to new host "https://n06a-eu.rcs.revma.com/an1ugyygzk8uv?rj-ttl=5&rj-tok=AAABhk0uEF4AhKT34DnVwteKSA"
audio_info:            buffers freed, free Heap: 80716 bytes
audio_info:            Connect to new host: "https://n06a-eu.rcs.revma.com/an1ugyygzk8uv?rj-ttl=5&rj-tok=AAABhk0uEF4AhKT34DnVwteKSA"
audio_info:            Connect to "n06a-eu.rcs.revma.com" on port 443, extension "/an1ugyygzk8uv?rj-ttl=5&rj-tok=AAABhk0uEF4AhKT34DnVwteKSA"
audio_info:            SSL has been established in 491 ms, free Heap: 34292 bytes
audio_info:            HTTP/1.0 200 OK
audio_info:            content-type: audio/mpeg, format is mp3
audio_info:            MP3Decoder has been initialized, free Heap: 36092 bytes
audio_info:            icy-name: an1ugyygzk8uv
audio_showstation:     an1ugyygzk8uv
audio_info:            Switch to DATA, metaint is 16000
audio_info:            stream ready
audio_info:            buffer filled in 15 ms
audio_info:            syncword found at pos 0
audio_info:            MP3 decode error -2 : MAINDATA_UNDERFLOW
audio_info:            syncword found at pos 0
audio_info:            Channels: 2
audio_info:            SampleRate: 44100
audio_info:            BitsPerSample: 16
audio_info:            BitRate: 128000
audio_info:            StreamTitle='Radio 357 - Najlepszy radiowy adres na świecie!'
audio_showstreamtitle: Radio 357 - Najlepszy radiowy adres na świecie!
audio_info:            slow stream, dropouts are possible
audio_info:            slow stream, dropouts are possible
```

And OK project.

```
audio_info:            Connect to new host: "https://stream.rcs.revma.com/an1ugyygzk8uv"
audio_info:            PSRAM found, inputBufferSize: 283615 bytes
audio_info:            buffers freed, free Heap: 80052 bytes
audio_info:            SSL has been established in 570 ms, free Heap: 35792 bytes
Setup completed.
audio_info:            redirect to new host "https://n05a-eu.rcs.revma.com/an1ugyygzk8uv?rj-ttl=5&rj-tok=AAABhk0km0QAg9dNl0_-BQtAnw"
audio_info:            Connect to new host: "https://n05a-eu.rcs.revma.com/an1ugyygzk8uv?rj-ttl=5&rj-tok=AAABhk0km0QAg9dNl0_-BQtAnw"
audio_info:            buffers freed, free Heap: 79372 bytes
audio_info:            SSL has been established in 525 ms, free Heap: 35408 bytes
audio_info:            icy-name: an1ugyygzk8uv
audio_showstation:     an1ugyygzk8uv
audio_info:            MP3Decoder has been initialized, free Heap: 35852 bytes
audio_info:            stream ready
audio_info:            syncword found at pos 0
audio_info:            Channels: 2
audio_info:            SampleRate: 44100
audio_info:            BitsPerSample: 16
audio_info:            BitRate: 128000
audio_info:            StreamTitle='Radio 357 - Najlepszy radiowy adres na świecie!'
audio_showstreamtitle: Radio 357 - Najlepszy radiowy adres na świecie!

```

**Comment by probonopd** [comment]
In my case, it seems that this helps (a bit):

```
    int freeInternal = heap_caps_get_free_size(MALLOC_CAP_INTERNAL);
    int freeExternal = heap_caps_get_free_size(MALLOC_CAP_SPIRAM);

    if (freeExternal > 0) {
        println("PSRAM detected");
    } else {
        println("PSRAM not detected");
        // Increase the buffer size of the built-in RAM buffer; this seem to help against stuttering
        println("Increasing buffer size");
        audio.setBufsize(15000, 0);
    }
```

**Comment by pdurys** [comment]
Just for reference. I guess my problem with dropouts are related to reported Issue with GPIO 0 https://github.com/schreibfaul1/ESP32-audioI2S/issues/408#issue-1438366058
Seems to be resolved with latest version 3.0.0
